halo2_proofs/src/dev/cost_model.rs

```diff
@@ -0,0 +1,364 @@
+    type Err = ParseIntError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Permutation {
+            columns: s.parse()?,
+        })
+    }
+}
+
+impl Permutation {
+    fn required_degree(&self) -> usize {
+        cmp::max(self.columns + 1, 2)
+    }
+
+    fn queries(&self) -> impl Iterator<Item = Poly> {
+        // - product commitments at x and x_inv
+        // - polynomial commitments at x
+        let product = "0,-1".parse().unwrap();
+        let poly = "0".parse().unwrap();
+
+        iter::empty()
+            .chain(Some(product))
+            .chain(iter::repeat(poly).take(self.columns))
+    }
+}
+
+/// High-level specifications of an abstract circuit.
+#[derive(Debug)]
+pub struct ModelCircuit {
+    /// Power-of-2 bound on the number of rows in the circuit.
+    pub k: usize,
+    /// Maximum degree of the circuit.
+    pub max_deg: usize,
+    /// Number of advice columns.
+    pub advice_columns: usize,
+    /// Number of lookup arguments.
+    pub lookups: usize,
+    /// Equality constraint permutation arguments.
+    pub permutations: Vec<Permutation>,
+    /// Number of distinct column queries across all gates.
+    pub column_queries: usize,
+    /// Number of distinct sets of points in the multiopening argument.
+    pub point_sets: usize,
+}
+
+impl From<CostOptions> for ModelCircuit {
+    fn from(opts: CostOptions) -> Self {
+        let max_deg = [1, opts.gate_degree]
+            .iter()
+            .cloned()
+            .chain(opts.lookup.iter().map(|l| l.required_degree()))
+            .chain(opts.permutation.iter().map(|p| p.required_degree()))
+            .max()
+            .unwrap();
+
+        let mut queries: Vec<_> = iter::empty()
+            .chain(opts.advice.iter())
+            .chain(opts.instance.iter())
+            .chain(opts.fixed.iter())
+            .cloned()
+            .chain(opts.lookup.iter().flat_map(|l| l.queries()))
+            .chain(opts.permutation.iter().flat_map(|p| p.queries()))
+            .chain(iter::repeat("0".parse().unwrap()).take(max_deg - 1))
+            .collect();
+
+        let column_queries = queries.len();
+        queries.sort_unstable();
+        queries.dedup();
+        let point_sets = queries.len();
+
+        ModelCircuit {
+            k: opts.k,
+            max_deg,
+            advice_columns: opts.advice.len(),
+            lookups: opts.lookup.len(),
+            permutations: opts.permutation,
+            column_queries,
+            point_sets,
+        }
+    }
+}
+
+impl ModelCircuit {
+    /// Size of the proof in bytes
+    pub fn proof_size<const COMM: usize, const SCALAR: usize>(
+        &self,
+        comm_scheme: CommitmentScheme,
+    ) -> usize {
+        let size = |points: usize, scalars: usize| points * COMM + scalars * SCALAR;
+
+        // PLONK:
+        // - COMM bytes (commitment) per advice column
+        // - 3 * COMM bytes (commitments) + 5 * SCALAR bytes (evals) per lookup argument
```

Comment:
```suggestion
        // - 3 * COMM bytes (commitments) + 5 * SCALAR bytes (evals) per lookup column
```

I think you pay for each of the columns rather than the argument itself.